### PR TITLE
Communication with SdI - sending invoice

### PR DESCRIPTION
**Pull request by torrocus**
## Pull Request Summary

We have implemented functionalities for sending invoices to SdI.

We've also introduced context, following the pattern established in other packages.

Furthermore, we've standardized CA (public) certificates, consolidating them into a single file. We have overcome authentication problems (on the client side).

## Feedback

We have removed the server-related part from this PR to complete the process of sending invoices to SdI as one whole.

Below are some logs from the console commands.

```console
$ go run ./cmd/gobl.fatturapa help
Usage:
  gobl.fatturapa [command]

Available Commands:
  completion  Generate the autocompletion script for the specified shell
  convert     Convert a GOBL JSON into a FatturaPA XML
  help        Help about any command
  server      Server for communication with SdI in selected environment
  transmit    Transmit a FatturaPA XML file to SdI in selected environment
  version     

Flags:
  -h, --help   help for gobl.fatturapa

Use "gobl.fatturapa [command] --help" for more information about a command.
```

### Transmit

```console
$ go run ./cmd/gobl.fatturapa help transmit
Transmit a FatturaPA XML file to SdI in selected environment

Usage:
  gobl.fatturapa transmit [file] [environment] [flags]

Flags:
      --ca-cert string   Path to a file containing the CA certificate
      --cert string      Path to a file containing the SDI PEM certificate
      --env string       Environment for running command (default "test")
  -h, --help             help for transmit
      --key string       Path to a file containing the sender PEM RSA private key
      --verbose          Logs all requests into the console
```

```console
$ go run ./cmd/gobl.fatturapa transmit --ca-cert ./ca-all.pem --cert ./SDI-B85905495.pem --key ./key_client.key IT01234567890_FPA01.xml 
Invoice sent to SdI.
Here is the response: {29218192 2024-05-31T13:54:56.491+02:00 <nil>}
```

----

Co-authored-by: @noplisu 
Co-authored-by: @torrocus

### CODE REVIEW

File: test/data/invoice-hotel-private.json
Comment (torrocus) [inline test/data/invoice-hotel-private.json:1]: We didn't touch this file. The change is probably due to rebase.

Comment (torrocus) [reply test/data/invoice-hotel-private.json:1]: This is exactly the change: ebf01b1a

I had to do a rebase on the fork and that's probably why.

File: sdi/server.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+		},
+		// func(request interface{}, w http.ResponseWriter, httpRequest *http.Request) (response interface{}, err error) {
+		nil,
+	)
+
+	// Interrupt signal handling
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
+
+	go func() {
```
Comment (samlown) [inline sdi/server.go:47]: This is a bit strange. Normally the waiting for a signal would happen after starting the HTTP server and subsequently request it to stop.

File: sdi/transport.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+type Transport interface {
+	SetTransport(t *http.Transport)
+	WithContext(ctx context.Context)
+	RoundTrip(r *http.Request) (*http.Response, error)
+}
+
+// DefaultTransport default configuration for HTTP communications
+type DefaultTransport struct {
+	HTTPTransport *http.Transport
+	Context       context.Context
```
Comment (samlown) [inline sdi/transport.go:20]: I wouldn't expect the context to be set inside something called "Transport". Context is based on the current execution context, not something that is long lived. As a general rule, the context should never be defined inside a struct as a variable, only passed on via method calls.

File: sdi/transport.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// SetTransport sets the underlying HTTPTransport value
+func (s *DefaultTransport) SetTransport(t *http.Transport) {
+	s.HTTPTransport = t
```
Comment (samlown) [inline sdi/transport.go:25]: I'd expect `HTTPTransport` to be a private attribute if it has a method to be configured.

File: sdi/transport.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+func (s *LoggingTransport) WithContext(ctx context.Context) {
+	s.Context = ctx
+}
+
+// RoundTrip executes a single HTTP transaction, returning,
+// a Response for the provided Request, but also logging that things for developer.
+func (s *LoggingTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	r := req.WithContext(s.Context)
+
+	bytes, _ := httputil.DumpRequestOut(r, true)
```
Comment (samlown) [inline sdi/transport.go:60]: Personally, I'd use `resty` or similar to make requests, as it has a lot of this functionality built in which is a bit easier to use.

File: test/data/invoice-hotel-private.json
Comment (samlown) [reply test/data/invoice-hotel-private.json:1]: If multiple people are working on a branch, I'd always recommend against rebase.

**Review by samlown — COMMENTED** [review]
In general the code looks good! But I understand that its still not working, which is frustrating :-(

File: test/data/invoice-hotel-private.json
Comment (torrocus) [reply test/data/invoice-hotel-private.json:1]: Is not so bad. Rebase is a great tool, but I agree that working on one branch can be more tedious. I recommend fixups and autosquash.

The problem was an outdated sdi branch.

File: sdi/transport.go (hunk 2 above)
Comment (torrocus) [reply sdi/transport.go]: TIL

File: sdi/helper.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+func ExtractXMLContent(data string, tag string) (string, error) {
+	openingTag := "<" + tag + ">"
+	closingTag := "</" + tag + ">"
+
+	openingTagIndex := strings.Index(data, openingTag)
+	if openingTagIndex == -1 {
+		return "", fmt.Errorf("opening tag <%s> not found", tag)
+	}
+
+	closingTagIndex := strings.Index(data, closingTag)
```
Comment (samlown) [inline sdi/helper.go:24]: You might want to use `LastIndex` to guarantee that you find the last entry.

File: sdi/helper.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	"fmt"
+	"mime"
+	"mime/multipart"
+	"strings"
+
+	resty "github.com/go-resty/resty/v2"
+)
+
+// ExtractXMLContent extracts the content inside specified XML tag.
+func ExtractXMLContent(data string, tag string) (string, error) {
```
Comment (samlown) [inline sdi/helper.go:15]: Do these methods need to be public?

File: sdi/helper.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	closingTagIndex += len(closingTag)
+
+	extractedXML := data[openingTagIndex:closingTagIndex]
+	return extractedXML, nil
+}
+
+// ParseMultipartResponse parses a multipart HTTP response and deserializes the content into the provided structure
+func ParseMultipartResponse(resp *resty.Response, response interface{}) error {
```
Comment (samlown) [inline sdi/helper.go:36]: This probably doesn't need to be a public method either.

File: sdi/invoice.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+	err = ParseMultipartResponse(resp, response)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// SoapRequestToSendInvoice prepares the request content for SOAP to send an invoice
+func SoapRequestToSendInvoice(fileName string, fileBody []byte) string {
```
Comment (samlown) [inline sdi/invoice.go:52]: public?

**Review by samlown — APPROVED** [review]
Looks good! Nice one! Just a few potential considerations with the public method names to consider.

File: sdi/helper.go (hunk 6 above)
Comment (torrocus) [reply sdi/helper.go]: I changed it to a private one, but I had to move it to the `sdi_test` package.

File: sdi/helper.go (hunk 7 above)
Comment (torrocus) [reply sdi/helper.go]: Good point, this function can be private.

File: sdi/invoice.go (hunk 8 above)
Comment (torrocus) [reply sdi/invoice.go:52]: In this case I have a bit of a dilemma. I know this function doesn't have to be public, but I like that it only does one thing and I can easily test it.
Any ideas?

**Pull request merged**